### PR TITLE
FPGA support

### PR DESCRIPTION
**Pull request by Deryugin**
* Initial implementation for fpga subsystem
* Add driver for socfpga
* Add `fpga` command to write FPGA configuration
* Add `hps_control_fpga_led` command which can cooperate with HPS_CONTROL_FPGA_LED.rbf config to control FPGA leds
* Update `arm/de0_nano_soc` template

I'll write how to run this example in corresponding wiki page, it's in progress.

~~This branch is still in progress because of couple issues:~~
* ~~There are some issues with caching in `hps_control_fpga_led` example~~
* ~~There are still stubs in fpga idesc ops~~

~~socfpga driver and fpga subsystem are mostly done, so it can be reviewed already :)~~

upd: Functional changes are done for now, so now all the code may be reviewed.

**Known issue:** `dd if=config of=/dev/fpga0` fails if `dd` uses too much chunks. I'm not going to fix it in this PR.

### CODE REVIEW

File: src/drivers/fpga/socfpga/socfpga.c [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+	log_debug(" GPIO_PORTA_EOI   %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_PORTA_EOI));
+	log_debug(" GPIO_EXT_PORTA   %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_EXT_PORTA));
+	log_debug(" GPIO_LS_SYNC     %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_LS_SYNC));
+	log_debug(" GPIO_VER_ID_CODE %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_VER_ID_CODE));
+	log_debug(" GPIO_CONFIG_REG2 %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_CONFIG_REG2));
+	log_debug(" GPIO_CONFIG_REG1 %08x", socfpga_read(socfpga, SOCFPGA_MGR_GPIO_CONFIG_REG1));
+}
+
+static uint32_t socfpga_mode(struct socfpga *socfpga) {
+	uint32_t ret = socfpga_read(socfpga,
```
Comment (anton-bondarev) [inline src/drivers/fpga/socfpga/socfpga.c]: May be separate a declaration and using?

File: src/drivers/fpga/socfpga/socfpga.c [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	if (err != 0) {
+		return err;
+	}
+
+	err = socfpga_mode_wait(socfpga, SOCFPGA_MODE_CONFIG);
+	if (err != 0) {
+		return -ETIMEDOUT;
+	}
+
+	/* Clear interrupt status */
```
Comment (anton-bondarev) [inline src/drivers/fpga/socfpga/socfpga.c]: just comment?

File: src/drivers/fpga/socfpga/socfpga.c [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+/**
+ * @brief Corresponds to Configuration Phase in docs
+ */
+static int socfpga_config_write(struct fpga *fpga, const uint8_t *buf, size_t len) {
+	struct socfpga *socfpga = fpga->priv;
+	uint32_t *buf32 = (uint32_t *) buf;
+	uint32_t tmp;
+
+	while (len > sizeof(uint32_t)) {
```
Comment (anton-bondarev) [inline src/drivers/fpga/socfpga/socfpga.c]: may be for will look better?

File: src/drivers/fpga/socfpga/socfpga.c [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+ */
+static int socfpga_config_complete(struct fpga *fpga) {
+	struct socfpga *socfpga = fpga->priv;
+	int err;
+	int timeout = TIMEOUT;
+	uint32_t tmp;
+
+	/* Wait for config write done */
+	while (timeout-- >= 0) {
+		static const uint32_t config_done =
```
Comment (anton-bondarev) [inline src/drivers/fpga/socfpga/socfpga.c]: separate declare and using?

File: src/drivers/fpga/socfpga/socfpga.c [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			SOCFPGA_GPIO_EXT_PORTA_NSP | SOCFPGA_GPIO_EXT_PORTA_CDP;
+		tmp = socfpga_read(socfpga, SOCFPGA_MGR_GPIO_EXT_PORTA);
+		if ((tmp & config_done) == config_done) {
```
Comment (anton-bondarev) [inline src/drivers/fpga/socfpga/socfpga.c]: maybe use macros directly?

File: src/drivers/fpga/socfpga/socfpga.c (hunk 2 above)
Comment (Deryugin) [reply src/drivers/fpga/socfpga/socfpga.c]: Out-of-date comment, removed.

File: src/drivers/fpga/socfpga/socfpga.c (hunk 5 above)
Comment (Deryugin) [reply src/drivers/fpga/socfpga/socfpga.c]: Fixed

File: src/drivers/fpga/socfpga/socfpga.c (hunk 4 above)
Comment (Deryugin) [reply src/drivers/fpga/socfpga/socfpga.c]: Removed variable

File: src/drivers/fpga/socfpga/socfpga.c (hunk 3 above)
Comment (Deryugin) [reply src/drivers/fpga/socfpga/socfpga.c]: I think that 4-byte alignment check will look uglier with for loop, and it doesn't really make big difference to use `for` or `while` here.

File: src/drivers/fpga/socfpga/socfpga.c (hunk 1 above)
Comment (Deryugin) [reply src/drivers/fpga/socfpga/socfpga.c]: I don't think that it makes much difference as we have a single variable here

**Pull request merged**